tests/src/whisk/core/database/test/MultipleReadersSingleWriterCacheTests.scala

```diff
@@ -0,0 +1,211 @@
+    for (i <- 1 to nIters) {
+      doCRUD("CONCURRENT CRUD to shared keys (long latency)", sharedKeys, 5000)
+        .nWriteInhibits.get should not be (0)
+    }
+  }
+
+  "the cache" should "support concurrent CRUD to shared keys, with update first" in {
+    for (i <- 1 to nIters) {
+      doCRUD("CONCURRENT CRUD to shared keys, with update first", sharedKeys, 1000, false)
+        .nWriteInhibits.get should be(0)
+    }
+  }
+
+  def sharedKeys = { i: Int => "foop_" + (i % 2) }
+
+  def doCRUD(
+    testName: String,
+    key: Int => String,
+    delay: Int = 1000,
+    readsFirst: Boolean = true,
+    nThreads: Int = 10): Inhibits = {
+
+    System.out.println();
+    System.out.println(testName);
+
+    val exec = Executors.newFixedThreadPool(nThreads)
+    val inhibits = Inhibits()
+
+    for (i <- 1 to nThreads) {
+      exec.submit(new Runnable { def run() = { new doReadWriteRead(key(i), inhibits, readsFirst).go(delay) } })
+    }
+
+    exec.shutdown
+    exec.awaitTermination(2, TimeUnit.MINUTES)
+
+    inhibits.debug
+    inhibits
+  }
+
+  case class Inhibits(
+      nReadInhibits: AtomicInteger = new AtomicInteger(0),
+      nWriteInhibits: AtomicInteger = new AtomicInteger(0)) {
+    def debug {
+      System.out.println("InhibitedReads: " + nReadInhibits);
+      System.out.println("InhibitedWrites: " + nWriteInhibits);
+    }
+  }
+
+  class doReadWriteRead(key: String, inhibits: Inhibits = Inhibits(), readFirst: Boolean = true) {
+    def go(implicit delay: Int): Inhibits = {
+      val latch = new CountDownLatch(2)
+
+      implicit val transId = TransactionId.testing
+
+      if (!readFirst) {
+        // we want to do the update before the first read
+        cacheUpdate(key, key, delayed("bar_b")).future onFailure {
+          case t =>
+            inhibits.nWriteInhibits.incrementAndGet();
+        }
+      }
+
+      System.out.println("R1");
+      cacheLookup(key, delayed("bar"), true) onComplete {
+        case Success(s) => {
+          latch.countDown()
+        }
+        case Failure(t) => {
+          latch.countDown()
+          inhibits.nReadInhibits.incrementAndGet();
+        }
+      }
+
+      if (readFirst) {
+        System.out.println("W");
+        // we did the read before the update, so do the write next
+        cacheUpdate(key, key, delayed("bar_b")).future onFailure {
+          case t =>
+            inhibits.nWriteInhibits.incrementAndGet();
+        }
+      }
+
+      System.out.println("R2");
+      cacheLookup(key, delayed("bar_c"), true) onComplete {
+        case Success(s) => {
+          latch.countDown();
+        }
+        case Failure(t) => {
+          inhibits.nReadInhibits.incrementAndGet();
+          latch.countDown();
+        }
+      }
+
+      latch.await(2, TimeUnit.MINUTES)
+      // System.out.println("DONE " + i);
+
+      inhibits
+    }
+  }
+
+  def delayed[W](v: W)(implicit delay: Int): Future[W] = {
```

Comment:
Should rather be `akka.after` instead of a blocking sleep. See http://doc.akka.io/docs/akka/2.4.10/scala/futures.html#After